src/service/history-impl.js

```diff
@@ -525,6 +545,28 @@ export class HistoryBindingNatural_ {
   }
 
   /**
+   * If this is a hash update the choice of `location.replace` vs
+   * `history.replaceState` is important. Due to bugs, not every browser
+   * triggers `:target` pseudo-class when `replaceState` is called.
+   * See http://www.zachleat.com/web/moving-target/ for more details.
+   * location.replace will trigger a `popstate` event, we temporarily
+   * disable handling it.
+   * @param {string} target
+   *
+   * @override
+   */
+  replaceStateForTarget(target) {
+    this.whenReady_(() => {
+      let hash = target.indexOf('#') == 0 ? target : `#${target}`;
+      this.ignoreUpcomingPopstate_ = true;
+      // TODO(mkhatib, #6095): Chrome iOS will add extra states for location.replace.
+      this.win.location.replace(hash);
+      this.ignoreUpcomingPopstate_ = false;
+      this.historyReplaceState_();
```

Comment:
Can we call this first? If so, we could drop the `ignoreUpcomingPopstate_` and instead check if the navigation is to the same URL as the previous. If it is, then we know there's not really a `popstate`.